features/cellular/framework/AT/AT_CellularContext.h

```diff
@@ -0,0 +1,124 @@
+    virtual void set_credentials(const char *apn, const char *uname = 0, const char *pwd = 0);
+    virtual const char *get_netmask();
+    virtual const char *get_gateway();
+
+private: // from CellularContext
+    virtual nsapi_error_t get_pdpcontext_params(pdpContextList_t &params_list);
+    virtual nsapi_error_t get_rate_control(CellularContext::RateControlExceptionReports &reports,
+                                           CellularContext::RateControlUplinkTimeUnit &time_unit, int &uplink_rate);
+    virtual nsapi_error_t get_apn_backoff_timer(int &backoff_timer);
+    virtual nsapi_error_t set_device_ready();
+    virtual nsapi_error_t set_sim_ready();
+    virtual nsapi_error_t register_to_network();
+    virtual nsapi_error_t attach_to_network();
+    virtual void set_file_handle(FileHandle *fh);
+
+protected:
+    virtual void cellular_callback(nsapi_event_t ev, intptr_t ptr);
+
+    /** Does the authentication for the PDP Context if user name and password are provided.
+     *  Can be overridden by the modem target if 3GPP default implementation if not an option
+     *
+     *  @return NSAPI_ERROR_OK if no credentials provided or authentication was successful
+     *          NSAPI_ERROR_AUTH_FAILURE if authentication failed
+     *          NSAPI_ERROR_DEVICE_ERROR if communication with the modemm failed
+     */
+    virtual nsapi_error_t do_user_authentication();
+
+    /** Activates PDP context or in PPP mode opens data channel.
+     *  Can be overridden by the modem target if 3GPP default implementation if not an option
+     */
+    virtual void do_connect();
+
+    /** Check if modem supports the given stack type. Can be overridden by the modem.
+     *
+     *  @return true if supported
+     */
+    virtual bool stack_type_supported(nsapi_ip_stack_t stack_type);
+
+    /** Get the operation specific timeout. Used in synchronous mode when setting the maximum
+     *   waiting time. Modem specific implementation can override this to provide different timeouts.
+     *
+     *  @param  op  current operation
+     *  @return     timeout in milliseconds
+     */
+    virtual uint32_t get_timeout_for_operation(ContextOperation op) const;
+
+    /** Helper method to call callback function if it is provided
+     *
+     *  @param status connection status which is parameter in callback function
+     */
+    void call_network_cb(nsapi_connection_status_t status);
+
+private:
+#if NSAPI_PPP_AVAILABLE
+    nsapi_error_t open_data_channel();
+    void ppp_status_cb(nsapi_event_t ev, intptr_t ptr);
+#endif // #if NSAPI_PPP_AVAILABLE
+    nsapi_error_t do_activate_context();
+    bool set_new_context(int cid);
+    bool get_context();
+    nsapi_error_t delete_current_context();
+    nsapi_ip_stack_t string_to_stack_type(const char *pdp_type);
+    nsapi_ip_stack_t get_stack_type();
+    nsapi_error_t check_operation(nsapi_error_t err, ContextOperation op);
+
+private:
+    nsapi_ip_stack_t _ip_stack_type_requested;
```

Comment:
Since there is no more requested stack type(removed from constructor) it means this is not relevant anymore.

Reply:
True, but we agreed that we are not remove this in this PR as it will change getting the context logic which will change anyway after this PR.